_posts/2018-08-09-gsoc-generic-servo.md

```diff
@@ -0,0 +1,99 @@
+
+## How it works
+
+The final idea of separation is based on using the [TypeHolderTrait](https://github.com/servo/servo/pull/21371/files#diff-12c9bc2e114f0cc3677b5a3b11349eaf). TypeHolderTrait is a trait with type parameters for every WebIDL interface that we want to extract from the script crate. *TypeHolderTrait* itself is defined in script crate with all type parameter traits. However, it is [implemented](https://github.com/servo/servo/pull/21371/files#diff-3092ed761ed8d9f72f5f0298b15b77d8R142) outside of the script crate so it can provide concrete types. TypeHolder enables us to use constructs like static methods in script crate. Later, we use TypeHolder as type parameter for structs and methods that require access to external types. Let's have origial dom struct like DOMParser:
+```rust
+struct DOMParser {
+    fn new_inherited(window: &Window) -> DOMParser {} 
+
+    fn new(window: &Window) -> DomRoot<DOMParser> {}
+
+    pub fn Constructor(window: &Window) -> Fallible<DomRoot<DOMParser>> {}
+}
+```
+
+This struct definition is removed from *script* crate. The DOMParserTrait with public methods is created and added as associated type to *TypeHolderTrait* which can than be used in place of original DOMParser type.
+
+```rust
+trait TypeHolderTrait {
+    type DomParser: DomParserTrait<Self>
+}
+
+
+trait DOMParserTrait<TH: TypeHolderTrait>: DOMParserMethods<TH> {
+    fn Constructor(window: &Window<TH>) -> Fallible<DomRoot<TH::DOMParser>>;
+}
+
+````
+
+## Effects on Servo
+
+Modifications in final PR should have only minimal effects on Servo speed. However, the codebase has undergone big surgery: over 12000 modified lines 🏥! The most intrusive change is to use TypeHolder where it is required. It turns out that TypeHolder is needed in a lot of places. Leading cause of such a large number of changes is mainly *GlobalScope* which is used all around the codebase and needs to be generic.
+
+Due to lack of Rust support for generic static variables at many places I had to modify the code to use initialization methods to fill static variables early during program initialization. For example, in Bindings, I have added an *InitTypeHolded* function which replaces the content of mutable static variables with appropriate function calls.
+
+## Speeeeed 🚀
+
+I have done testing on MacBook Pro 2015, High Sierra, 2,7 GHz i5 dual core, 16 GB RAM using `/usr/bin/time cargo build`, which shows maximal memory usage during compilation. Results may vary depending on build machine.
+
+Resources were measured in this way:
+
+1. compile full servo
+2. modify Attr.rs file and ServoParser.rs files so I force recompilatiion of crates. After separation Attr.rs and ServoParser.rs are in different crates.
```

Comment:
- "recompilation"
- This could still be more clear - does the "multiple script crates" time measurement only include touching ServoParser.rs? My measurements in https://github.com/servo/servo/pull/21371 separated the times for touching a file in the script crate from the times for touching a file in the script_servoparser crate.